### PR TITLE
fix: got status extension handle error

### PR DESCRIPTION
**Pull request by rarkins**

### CODE REVIEW

**Comment by rarkins** [comment]
@szmarczak does this change look right to you? It's causing a bunch of `nock`'d tests to fail, but I don't know if that necessarily means the change is wrong

File: lib/util/got/stats.ts [hunk 1]
```diff
@@ -42,17 +42,23 @@ export const printStats = (): void => {
...
-    const nextPromise = next(options);
-    nextPromise.on('response', () => {
-      delete outstandingRequests[request];
-      const elapsed = new Date().getTime() - start.getTime();
-      stats[options.hostname] = stats[options.hostname] || [];
-      stats[options.hostname].push(elapsed);
+    const promise = new Promise((resolve, reject) => {
+      const request = `${options.method} ${options.href}`;
+      logger.info(request);
+      outstandingRequests[request] = true;
```
Comment (rarkins) [inline lib/util/got/stats.ts]: Note: this object was found to have a request in it when Renovate "hangs". i.e. `.on('response` is not getting called.

**Comment by szmarczak** [comment]
Returning your own custom Promise is buggy in Got 9. I don't recommend doing that.

**Comment by szmarczak** [comment]
This PR seems to be pointless as it doesn't change anything (well it actually breaks some Got things).

File: lib/util/got/stats.ts [hunk 2]
```diff
@@ -42,17 +42,23 @@ export const printStats = (): void => {
...
+        stats[options.hostname] = stats[options.hostname] || [];
+        stats[options.hostname].push(elapsed);
+        resolve(nextPromise);
+      });
+      nextPromise.on('error', () => {
+        reject(nextPromise);
+      });
     });
-    return nextPromise;
+    return promise;
```
Comment (viceice) [inline lib/util/got/stats.ts]:
```suggestion
    const request = `${options.method} ${options.href}`;
    logger.info(request);
    outstandingRequests[request] = true;
    const start = new Date();
    const nextPromise = next(options);
    nextPromise.on('response', () => {
      delete outstandingRequests[request];
      const elapsed = new Date().getTime() - start.getTime();
      stats[options.hostname] = stats[options.hostname] || [];
      stats[options.hostname].push(elapsed);
    });
    nextPromise.on('error', () => {
      delete outstandingRequests[request];
      // TODO: count errors per host?
    });
    return nextPromise;
```

@szmarczak Can this work?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/got/stats.ts (hunk 2 above)
Comment (szmarczak) [reply lib/util/got/stats.ts]: Yup.

Comment (viceice) [reply lib/util/got/stats.ts]: @szmarczak I've looked in the got code, you never invoke `on('error', ...)`. Is that a bug?

Comment (szmarczak) [reply lib/util/got/stats.ts]: @viceice No, you actually need to call `nextPromise.catch(...)`.

**Pull request closed without merging**